src/microservices/Microservices.IsIdentifiable/Runners/IsIdentifiableAbstractRunner.cs

```diff
@@ -219,14 +240,41 @@ public void LoadRules(string yaml)
 
         // ReSharper disable once UnusedMemberInSuper.Global
         public abstract int Run();
-
+        
         /// <summary>
         /// Returns each subsection of <paramref name="fieldValue"/> which violates validation rules (e.g. the CHI found).
         /// </summary>
         /// <param name="fieldName"></param>
         /// <param name="fieldValue"></param>
         /// <returns></returns>
         protected IEnumerable<FailurePart> Validate(string fieldName, string fieldValue)
+        {
+            // every time we lookup a value increment the counter and see if we should check cache overflow
+            if(_valuesLookedUp++ % _checkEvery == 0)
+            {
+                _valuesLookedUp = 0;
+                foreach (var c in Caches)
+                {
+                    //any column with more than the maximum number of cached values should be cleared
+                    if(c.Value.Keys.Count > MaxCacheSize)
+                        c.Value.Clear();
+                }
+            }
+
+            // make sure that we have a cache for this column name
+            var cache = Caches.GetOrAdd(fieldName,(v)=>new ConcurrentDictionary<string, FailurePart[]>());
+
+            // lookup the cached value or call ValidateImpl
+            return cache.GetOrAdd(fieldValue?? "NULL",(k)=>ValidateImpl(fieldName,fieldValue).ToArray());
+        }
+        
+        /// <summary>
+        /// Actual implementation of <see cref="Validate(string, string)"/> after a cache miss has occurred.  This method is only called when a cached answer is not found for the given <paramref name="fieldName"/> and <paramref name="fieldValue"/> pair
+        /// </summary>
+        /// <param name="fieldName"></param>
+        /// <param name="fieldValue"></param>
+        /// <returns></returns>
+        protected IEnumerable<FailurePart> ValidateImpl(string fieldName, string fieldValue)
         {
             if (_skipColumns.Contains(fieldName))
```

Comment:
Should this check (and any others) be pulled up from the Impl method so we can bail out early before dealing with the cache code?

Reply:
I think we should keep all the logic in `ValidateImpl` that way if somoene overiddes it later on they don't have to worry about making changes in 2 places.  The caching always happens first and always caches only what your `ValidateImpl` method returned.